Switch to `EarlyBinder` for `type_of` query

kylematsuda opened the pull request:
Part of the work to finish #105779 and implement https://github.com/rust-lang/types-team/issues/78.

Several queries `X` have a `bound_X` variant that wraps the output in `EarlyBinder`. This adds `EarlyBinder` to the return type of the `type_of` query and removes `bound_type_of`.

r? @lcnr

kylematsuda committed:
change usages of type_of to bound_type_of

kylematsuda committed:
remove bound_type_of query; make type_of return EarlyBinder; change type_of in metadata

kylematsuda commented on compiler/rustc_borrowck/src/diagnostics/region_errors.rs:
```diff
@@ -575,7 +575,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
         let mut output_ty = self.regioncx.universal_regions().unnormalized_output_ty;
         if let ty::Alias(ty::Opaque, ty::AliasTy { def_id, .. }) = *output_ty.kind() {
-            output_ty = self.infcx.tcx.type_of(def_id)
+            output_ty = self.infcx.tcx.bound_type_of(def_id).subst_identity()
```
use `substs` from `AliasTy`?

kylematsuda commented on compiler/rustc_borrowck/src/diagnostics/region_errors.rs:
```diff
@@ -896,7 +896,7 @@ impl<'a, 'tcx> MirBorrowckCtxt<'a, 'tcx> {
 
             // Only suggest this on function calls, not closures
-            let ty = tcx.type_of(fn_did);
+            let ty = tcx.bound_type_of(fn_did).subst_identity();
```
use `substs` from `FnDef` above?

kylematsuda commented on compiler/rustc_borrowck/src/universal_regions.rs:
```diff
@@ -674,7 +674,7 @@ impl<'cx, 'tcx> UniversalRegionsBuilder<'cx, 'tcx> {
                 // "output" (the type of the constant).
                 assert_eq!(self.mir_def.did.to_def_id(), def_id);
-                let ty = tcx.type_of(self.mir_def.def_id_for_type_of());
+                let ty = tcx.bound_type_of(self.mir_def.def_id_for_type_of()).subst_identity();
```
use `substs` from `DefiningTy::Const`? also `DefiningTy::FnDef` a few lines above also has `substs` that could be used on the `fn_sig`

kylematsuda commented on compiler/rustc_hir_analysis/src/variance/constraints.rs:
```diff
@@ -112,7 +112,7 @@ impl<'a, 'tcx> ConstraintContext<'a, 'tcx> {
                     self.add_constraints_from_ty(
                         current_item,
-                        tcx.type_of(field.did),
+                        tcx.bound_type_of(field.did).subst_identity(),
```
use `substs` from `Adt`?

kylematsuda commented on compiler/rustc_hir_typeck/src/method/suggest.rs:
```diff
@@ -906,8 +906,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                         // So we avoid suggestion method with Box<Self>
                                         // for instance
-                                        self.tcx.at(span).type_of(*def_id) != rcvr_ty
-                                            && self.tcx.at(span).type_of(*def_id) != rcvr_ty
```
This is sort of unrelated to adding `EarlyBinder`, but I was confused by this... what's going on here?

BoxyUwU replied:
yeah this is really odd lol I have no idea why its checking the same condition twice...

kylematsuda replied:
Hi, so I did a little digging on this and tracked it down to https://github.com/rust-lang/rust/pull/104422 (which renamed `actual` to `rcvr_ty`, whereas originally both names were in scope). I might try to think of a test case where the original check is needed

kylematsuda commented on compiler/rustc_lint/src/builtin.rs:
```diff
@@ -2853,8 +2859,8 @@ impl ClashingExternDeclarations {
-                                        tcx.type_of(a_did),
-                                        tcx.type_of(b_did),
+                                        tcx.bound_type_of(a_did).subst_identity(),
+                                        tcx.bound_type_of(b_did).subst_identity(),
```
it seems like there are possible `substs` here from the `(Adt(..), Adt(..))` above

kylematsuda commented on compiler/rustc_middle/src/ty/print/pretty.rs:
```diff
@@ -755,7 +755,7 @@ pub trait PrettyPrinter<'tcx>:
                         // `type_of` on a type alias or assoc type should never cause a cycle.
                         if let ty::Alias(ty::Opaque, ty::AliasTy { def_id: d, .. }) =
-                            *self.tcx().type_of(parent).kind()
+                            *self.tcx().bound_type_of(parent).subst_identity().kind()
```
use `substs` from `AliasTy`?

kylematsuda commented on compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs:
```diff
@@ -695,7 +695,7 @@ fn transform_ty<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>, options: TransformTyOptio
                 let param_env = tcx.param_env(variant.def_id);
                 let field = variant.fields.iter().find(|field| {
-                    let ty = tcx.type_of(field.did);
+                    let ty = tcx.bound_type_of(field.did).subst_identity();
```
use `substs` from `Adt` here?

kylematsuda reviewed (commented):
Here are the places where I thought there were potential `substs` but wasn't sure whether they should be used

bors commented:
:umbrella: The latest upstream changes (presumably #107768) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

BoxyUwU commented:
r? @BoxyUwU

lcnr commented:
handing this off to @BoxyUwU as I don't have a lot of time atm ^^

I realized is that the reason I didn't see your comments in your previous PR was that they are for the commit which uses `bound_X` everywhere and I tend to only review the the final state of the PR. For future PRs it is probably better to only comment on the final changes as dealing with comments on specific diffs isn't great.

Thanks for your work :heart:

bors commented:
:umbrella: The latest upstream changes (presumably #107924) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

BoxyUwU commented on compiler/rustc_hir_typeck/src/demand.rs:
```diff
@@ -2036,7 +2036,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                                    kind: CallableKind| {
             let arg_idx = args.iter().position(|a| a.hir_id == expr.hir_id).unwrap();
-            let fn_ty = self.tcx.bound_type_of(def_id).0;
+            let fn_ty = self.tcx.type_of(def_id).0;
```
```suggestion
            let fn_ty = self.tcx.type_of(def_id).skip_binder();
```

BoxyUwU commented on compiler/rustc_hir_typeck/src/fn_ctxt/_impl.rs:
```diff
@@ -1225,7 +1225,13 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                        self.fcx.ct_infer(tcx.type_of(param.def_id), Some(param), inf.span).into()
+                        self.fcx
+                            .ct_infer(
+                                tcx.type_of(param.def_id).subst_identity(),
```
```suggestion
                                tcx.type_of(param.def_id).no_bound_vars().expect("const parameter types cannot be generic"),
```

BoxyUwU commented on compiler/rustc_hir_typeck/src/method/confirm.rs:
```diff
@@ -384,7 +384,13 @@ impl<'a, 'tcx> ConfirmContext<'a, 'tcx> {
-                        self.cfcx.ct_infer(tcx.type_of(param.def_id), Some(param), inf.span).into()
+                        self.cfcx
+                            .ct_infer(
+                                tcx.type_of(param.def_id).subst_identity(),
```
```suggestion
                                tcx.type_of(param.def_id).no_bound_vars().expect("const parameter types cannot be generic"),
```

BoxyUwU commented on compiler/rustc_hir_typeck/src/method/probe.rs:
```diff
@@ -1944,7 +1944,7 @@ impl<'a, 'tcx> ProbeContext<'a, 'tcx> {
                     span,
                 };
-                self.next_const_var(self.tcx.type_of(param.def_id), origin).into()
+                self.next_const_var(self.tcx.type_of(param.def_id).subst_identity(), origin).into()
```
```suggestion
                self.next_const_var(self.tcx.type_of(param.def_id).no_bound_vars().expect("const parameter types cannot be generic"), origin).into()
```

BoxyUwU commented on compiler/rustc_infer/src/infer/mod.rs:
```diff
@@ -1175,7 +1175,9 @@ impl<'tcx> InferCtxt<'tcx> {
                     });
-                self.tcx.mk_const(const_var_id, self.tcx.type_of(param.def_id)).into()
+                self.tcx
+                    .mk_const(const_var_id, self.tcx.type_of(param.def_id).subst_identity())
```
```suggestion
                    .mk_const(const_var_id, self.tcx.type_of(param.def_id).no_bound_vars().expect("const parameter types cannot be generic"))
```

BoxyUwU commented on compiler/rustc_middle/src/mir/mod.rs:
```diff
@@ -2483,7 +2483,7 @@ impl<'tcx> ConstantKind<'tcx> {
         debug!("expr.kind: {:?}", expr.kind);
 
-        let ty = tcx.type_of(def.def_id_for_type_of());
+        let ty = tcx.type_of(def.def_id_for_type_of()).subst_identity();
```
```suggestion
        let ty = tcx.type_of(def.def_id_for_type_of()).no_bound_vars().expect("const parameter types cannot be generic");
```

BoxyUwU commented on compiler/rustc_trait_selection/src/traits/select/confirmation.rs:
```diff
@@ -557,7 +557,7 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
                                         ty::BoundVar::from_usize(bound_vars.len() - 1),
                                     ),
-                                    tcx.type_of(param.def_id),
+                                    tcx.type_of(param.def_id).subst_identity(),
```
```suggestion
                                    tcx.type_of(param.def_id).no_bound_vars().expect("const parameter types cannot be generic"),
```

kylematsuda commented on compiler/rustc_codegen_llvm/src/debuginfo/mod.rs:
```diff
@@ -508,7 +508,7 @@ impl<'ll, 'tcx> DebugInfoMethods<'tcx> for CodegenCx<'ll, 'tcx> {
                     let impl_self_ty = cx.tcx.subst_and_normalize_erasing_regions(
```
Okay! Yeah I figured that should probably be changed. Okay to open a follow-up PR fixing that after this one?

BoxyUwU replied:
Yea thats fine ^^

kylematsuda committed:
changes from review

kylematsuda committed:
fix new usage of type_of

kylematsuda commented:
Hi @BoxyUwU, thanks for reviewing! I implemented all of those changes in the 2nd to last commit; the last commit is just a small change that came up when I rebased :smiley:

bors commented:
:umbrella: The latest upstream changes (presumably #108079) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #108006) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:umbrella: The latest upstream changes (presumably #108096) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

BoxyUwU commented:
@bors r+ p=1 (because this PR seems to accumulate merge conflicts like something that accumulates lots of stuff)

I think the stuff about using in scope substs instead of `subst_identity` is probably not super urgent :thinking: having `type_of` return `EarlyBinder` in newly written code feels like a good step even if every single existing `type_of` call in the codebase hasn't been audited for "uses correct substs" :sparkles:

bors commented:
:pushpin: Commit f6c3469aa2f445c0eb4a7598175cba969ef4b948 has been approved by `BoxyUwU`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=BoxyUwU f6c3469aa2f445c0eb4a7598175cba969ef4b948 -->
<!-- homu: {"type":"Approved","sha":"f6c3469aa2f445c0eb4a7598175cba969ef4b948","approver":"BoxyUwU","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit f6c3469aa2f445c0eb4a7598175cba969ef4b948 with merge 9556b56dbdbd4238f0051e7230004b0aa488fa14...
<!-- homu: {"type":"BuildStarted","head_sha":"f6c3469aa2f445c0eb4a7598175cba969ef4b948","merge_sha":"9556b56dbdbd4238f0051e7230004b0aa488fa14"} -->

bors commented:
:umbrella: The latest upstream changes (presumably #108145) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/4200771887/jobs/7289420413)
Approved by: BoxyUwU
Pushing 9556b56dbdbd4238f0051e7230004b0aa488fa14 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"BoxyUwU","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/4200771887/jobs/7289420413"},"merge_sha":"9556b56dbdbd4238f0051e7230004b0aa488fa14"} -->

rust-timer commented:
Finished benchmarking commit (9556b56dbdbd4238f0051e7230004b0aa488fa14): [comparison URL](https://perf.rust-lang.org/compare.html?start=ea218392a4ce119c4dfcd8fb94a7fee77f76f2c5&end=9556b56dbdbd4238f0051e7230004b0aa488fa14&stat=instructions:u).

### Overall result: ✅ improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=ea218392a4ce119c4dfcd8fb94a7fee77f76f2c5&end=9556b56dbdbd4238f0051e7230004b0aa488fa14&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-2.4%, -2.4%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=ea218392a4ce119c4dfcd8fb94a7fee77f76f2c5&end=9556b56dbdbd4238f0051e7230004b0aa488fa14&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=ea218392a4ce119c4dfcd8fb94a7fee77f76f2c5&end=9556b56dbdbd4238f0051e7230004b0aa488fa14&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -2.4% | [-3.9%, -1.2%] | 9     |
| All ❌✅ (primary)                 | -     | -              | 0     |
</details>
